File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/BiologicalSimulationLaboratory.java
package com.gtladd.gtladditions.common.machine.muiltblock.controller;

import org.gtlcore.gtlcore.api.machine.multiblock.ParallelMachine;
import org.gtlcore.gtlcore.common.machine.multiblock.electric.StorageMachine;
import org.gtlcore.gtlcore.utils.Registries;

import com.gregtechceu.gtceu.api.capability.recipe.IO;
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.IRecipeLogicMachine;
import com.gregtechceu.gtceu.api.machine.multiblock.WorkableElectricMultiblockMachine;
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic;
import com.gregtechceu.gtceu.api.recipe.GTRecipe;
import com.gregtechceu.gtceu.api.recipe.RecipeHelper;
import com.gregtechceu.gtceu.utils.FormattingUtil;

import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

import com.gtladd.gtladditions.api.recipeslogic.GTLAddMultipleRecipesLogic;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

public class BiologicalSimulationLaboratory extends StorageMachine implements ParallelMachine {

    private static double reDuctionEUt = 1.0;
    private static double reDuctionDuration = 1.0;
    private static int Max_Parallels = 64;
    private static boolean Is_MultiRecipe = false;
    private static final ItemStack RHENIUM_NANOSWARM = Registries.getItemStack("gtceu:rhenium_nanoswarm");
    private static final ItemStack ORICHALCUM_NANOSWARM = Registries.getItemStack("gtceu:orichalcum_nanoswarm");
    private static final ItemStack INFUSCOLIUM_NANOSWARM = Registries.getItemStack("gtceu:infuscolium_nanoswarm");
    private static final ItemStack NAN_CERTIFICATE = Registries.getItemStack("gtceu:nan_certificate");

    public BiologicalSimulationLaboratory(IMachineBlockEntity holder) {
        super(holder, 1);
    }

    protected @NotNull RecipeLogic createRecipeLogic(Object... args) {
        return new BiologicalSimulationLaboratoryLogic(this);
    }

    protected boolean filter(@NotNull ItemStack itemStack) {
        Item item = itemStack.getItem();
        if (RHENIUM_NANOSWARM.is(item)) return true;
        else if (ORICHALCUM_NANOSWARM.is(item)) return true;
        else if (INFUSCOLIUM_NANOSWARM.is(item)) return true;
        else return NAN_CERTIFICATE.is(item);
    }

    public static boolean beforeWorking(IRecipeLogicMachine machine, @NotNull GTRecipe recipe) {
        if (machine instanceof BiologicalSimulationLaboratory biologicalSimulationLaboratory) {
            biologicalSimulationLaboratory.setparameter((MetaMachine) machine);
            List<ItemStack> input = RecipeHelper.getInputItems(recipe);
            for (ItemStack itemstack : input) {
                if (itemstack.getItem().equals(Registries.getItem("avaritia:infinity_sword")) && !Is_MultiRecipe) return false;
            }
        }
        return true;
    }

    public void addDisplayText(@NotNull List<Component> textList) {
        super.addDisplayText(textList);
        if (this.isFormed) {
            if (this.holder.getOffsetTimer() % 20L == 0L) {
                this.setparameter(this);
            }
            textList.add(Component.translatable("gtceu.multiblock.parallel", Component.translatable(FormattingUtil.formatNumbers(Max_Parallels)).withStyle(ChatFormatting.DARK_PURPLE)).withStyle(ChatFormatting.GRAY));
            textList.add(Component.translatable((Is_MultiRecipe ? "已" : "未") + "解锁寰宇支配之剑的配方"));
            textList.add(Component.translatable("gtceu.machine.eut_multiplier.tooltip", Component.translatable(FormattingUtil.formatNumbers(reDuctionEUt))));
            textList.add(Component.translatable("gtceu.machine.duration_multiplier.tooltip", Component.translatable(FormattingUtil.formatNumbers(reDuctionDuration))));
        }
    }

    private int getTier(MetaMachine machine) {
        if (machine instanceof BiologicalSimulationLaboratory biologicalSimulationLaboratory) {
            Item item = biologicalSimulationLaboratory.machineStorage.storage.getStackInSlot(0).getItem();
            if (RHENIUM_NANOSWARM.is(item)) return 1;
            else if (ORICHALCUM_NANOSWARM.is(item)) return 2;
            else if (INFUSCOLIUM_NANOSWARM.is(item)) return 3;
            else if (NAN_CERTIFICATE.is(item)) return 4;
        }
        return 0;
    }

    private void setparameter(MetaMachine machine) {
        int tier = getTier(machine);
        switch (tier) {
            case 1 -> setMachine(false, 2048, 0.9, 0.9);
            case 2 -> setMachine(false, 16384, 0.8, 0.6);
            case 3 -> setMachine(false, 262144, 0.6, 0.4);
            case 4 -> setMachine(true, 4194304, 0.25, 0.1);
            default -> setMachine(false, 64, 1.0, 1.0);
        }
    }

    private void setMachine(boolean isMultiRecipe, int maxParallel, double Reductioneut, double Reductionduration) {
        Is_MultiRecipe = isMultiRecipe;
        Max_Parallels = maxParallel;
        reDuctionEUt = Reductioneut;
        reDuctionDuration = Reductionduration;
    }

    @Override
    public int getMaxParallel() {
        return Max_Parallels;
    }

    private static class BiologicalSimulationLaboratoryLogic extends GTLAddMultipleRecipesLogic {

        public BiologicalSimulationLaboratoryLogic(WorkableElectricMultiblockMachine machine) {
            super((ParallelMachine) machine);
        }

        public BiologicalSimulationLaboratory getMachine() {
            return (BiologicalSimulationLaboratory) super.getMachine();
        }

        public void findAndHandleRecipe() {
            lastRecipe = null;
            GTRecipe match;
            if (isNanCertificate()) match = getRecipe();
            else match = getOneRecipe();
            if (match != null && match.matchRecipe(this.machine).isSuccess()) {
                setupRecipe(match);
            }
        }

        private boolean isNanCertificate() {
            ItemStack item = getMachine().getMachineStorageItem();
            return Objects.equals(item.getItem(), Registries.getItem("gtceu:nan_certificate"));
        }

        private GTRecipe getOneRecipe() {
            if (!machine.hasProxies()) return null;
            GTRecipe recipe = this.machine.getRecipeType().getLookup().findRecipe(machine);
            if (recipe == null || RecipeHelper.getRecipeEUtTier(recipe) > getMachine().getTier()) return null;
            recipe = parallelRecipe(recipe, getMachine().getMaxParallel());
            RecipeHelper.setInputEUt(recipe, (long) Math.max(1.0, (RecipeHelper.getInputEUt(recipe) * reDuctionEUt)));
            recipe.duration = (int) Math.max(1.0, (double) recipe.duration *
                    reDuctionDuration / (1 << (getMachine().getTier() - RecipeHelper.getRecipeEUtTier(recipe))));
            return recipe;
        }

        @Override
        public void onRecipeFinish() {
            machine.afterWorking();
            if (lastRecipe != null) {
                lastRecipe.postWorking(this.machine);
                lastRecipe.handleRecipeIO(IO.OUT, this.machine, this.chanceCaches);
            }
            GTRecipe match = isNanCertificate() ? getRecipe() : getOneRecipe();
            if (match != null) if (match.matchRecipe(this.machine).isSuccess()) {
                setupRecipe(match);
                return;
            }
            setStatus(Status.IDLE);
            progress = 0;
            duration = 0;
        }
    }
}

File: src/main/java/com/gtladd/gtladditions/common/machine/muiltblock/controller/BiologicalSimulationLaboratory.kt
package com.gtladd.gtladditions.common.machine.muiltblock.controller

import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.gui.GuiTextures
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.feature.IMachineLife
import com.gregtechceu.gtceu.api.machine.feature.IRecipeLogicMachine
import com.gregtechceu.gtceu.api.machine.trait.NotifiableItemStackHandler
import com.gregtechceu.gtceu.api.machine.trait.RecipeLogic
import com.gregtechceu.gtceu.api.recipe.GTRecipe
import com.gregtechceu.gtceu.api.recipe.RecipeHelper
import com.gregtechceu.gtceu.api.recipe.modifier.ParallelLogic
import com.gregtechceu.gtceu.common.data.GTItems
import com.gregtechceu.gtceu.utils.FormattingUtil
import com.gtladd.gtladditions.api.machine.logic.GTLAddMultipleRecipesLogic
import com.gtladd.gtladditions.api.machine.multiblock.GTLAddWorkableElectricMultipleRecipesMachine
import com.lowdragmc.lowdraglib.gui.widget.SlotWidget
import com.lowdragmc.lowdraglib.gui.widget.Widget
import com.lowdragmc.lowdraglib.gui.widget.WidgetGroup
import com.lowdragmc.lowdraglib.misc.ItemStackTransfer
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import net.minecraft.network.chat.Component
import net.minecraft.world.item.ItemStack
import org.gtlcore.gtlcore.api.recipe.RecipeResult
import org.gtlcore.gtlcore.api.recipe.RecipeRunnerHelper.handleRecipeOutput
import org.gtlcore.gtlcore.api.recipe.RecipeRunnerHelper.matchRecipeOutput
import org.gtlcore.gtlcore.common.data.GTLRecipeModifiers
import org.gtlcore.gtlcore.utils.Registries.getItem
import org.gtlcore.gtlcore.utils.Registries.getItemStack
import java.util.function.BiPredicate

class BiologicalSimulationLaboratory(holder: IMachineBlockEntity) :
    GTLAddWorkableElectricMultipleRecipesMachine(holder), IMachineLife {

    @field:Persisted
    val machineStorage: NotifiableItemStackHandler = createMachineStorage()

    private var reductionEUt = 1.0
    private var reductionDuration = 1.0
    private var maxParallels = 64
    private var isMultiRecipe = false

    fun createMachineStorage(): NotifiableItemStackHandler {
        val handler = NotifiableItemStackHandler(this, 1, IO.NONE, IO.BOTH) { slots ->
            object : ItemStackTransfer(1) {
                override fun getSlotLimit(slot: Int): Int = 1
                override fun onContentsChanged() {
                    recalculateParameters()
                }
            }
        }
        handler.setFilter { itemStack: ItemStack? -> this.filter(itemStack!!) }
        return handler
    }

    override fun createRecipeLogic(vararg args: Any): RecipeLogic {
        return BiologicalSimulationLaboratoryLogic(this)
    }

    override fun getRecipeLogic(): BiologicalSimulationLaboratoryLogic {
        return super.getRecipeLogic() as BiologicalSimulationLaboratoryLogic
    }

    override fun onStructureFormed() {
        super.onStructureFormed()
        recalculateParameters()
    }

    fun filter(itemStack: ItemStack): Boolean {
        val item = itemStack.item
        return NAN_CERTIFICATE.`is`(item) || INFUSCOLIUM_NANOSWARM.`is`(item) || ORICHALCUM_NANOSWARM.`is`(item) || RHENIUM_NANOSWARM.`is`(
            item
        )
    }

    fun isMultiRecipeEnabled(): Boolean = isMultiRecipe

    override fun createUIWidget(): Widget {
        val widget = super.createUIWidget()
        if (widget is WidgetGroup) {
            val size = widget.size
            widget.addWidget(
                SlotWidget(machineStorage.storage, 0, size.width - 30, size.height - 30, true, true)
                    .setBackground(GuiTextures.SLOT)
            )
        }
        return widget
    }

    override fun addDisplayText(textList: MutableList<Component?>) {
        super.addDisplayText(textList)
        if (this.isFormed) {
            textList.add(Component.translatable("gtceu.machine.biological_simulation_laboratory.gui.tooltip." + if (isMultiRecipe) 1 else 0))
            textList.add(
                Component.translatable(
                    "gtceu.machine.eut_multiplier.tooltip", Component.translatable(
                        FormattingUtil.formatNumbers(reductionEUt))
                )
            )
            textList.add(
                Component.translatable(
                    "gtceu.machine.duration_multiplier.tooltip", Component.translatable(
                        FormattingUtil.formatNumbers(reductionDuration))
                )
            )
        }
    }

    private fun recalculateParameters() {
        val item = machineStorage.storage.getStackInSlot(0).item
        when {
            RHENIUM_NANOSWARM.`is`(item) -> setMachineParameters(false, 2048, 0.9, 0.9)
            ORICHALCUM_NANOSWARM.`is`(item) -> setMachineParameters(false, 16384, 0.8, 0.6)
            INFUSCOLIUM_NANOSWARM.`is`(item) -> setMachineParameters(false, 262144, 0.6, 0.4)
            NAN_CERTIFICATE.`is`(item) -> setMachineParameters(true, 4194304, 0.25, 0.1)
            else -> setMachineParameters(false, 64, 1.0, 1.0)
        }
    }

    private fun setMachineParameters(isMultiRecipe: Boolean, maxParallel: Int, reductionEut: Double, reductionDuration: Double) {
        this@BiologicalSimulationLaboratory.isMultiRecipe = isMultiRecipe
        maxParallels = maxParallel
        reductionEUt = reductionEut
        this@BiologicalSimulationLaboratory.reductionDuration = reductionDuration
    }

    override fun onMachineRemoved() = clearInventory(machineStorage)

    override fun getMaxParallel(): Int = maxParallels

    override fun getFieldHolder(): ManagedFieldHolder = MANAGED_FIELD_HOLDER

    class BiologicalSimulationLaboratoryLogic(machine: BiologicalSimulationLaboratory) :
        GTLAddMultipleRecipesLogic(machine, BEFORE_RECIPE) {

        override fun getMachine(): BiologicalSimulationLaboratory {
            return super.getMachine() as BiologicalSimulationLaboratory
        }

        override fun findAndHandleRecipe() {
            lastRecipe = null
            recipeStatus = null
            val match = if (this.isNanCertificate) gtRecipe
            else this.oneRecipe
            if (match != null && matchRecipeOutput(this.machine, match)) {
                setupRecipe(match)
            }
        }

        val isNanCertificate: Boolean
            get() {
                val item = getMachine().machineStorage.storage.getStackInSlot(0)
                return item.item == getItem("gtceu:nan_certificate")
            }

        val oneRecipe: GTRecipe?
            get() {
                if (!machine.hasProxies()) return null
                val lab = getMachine()
                var recipe = machine.recipeType.lookup.find(machine, this::checkRecipe)
                if (recipe == null) return null
                recipe = ParallelLogic.applyParallel(lab, recipe,
                    parallel.maxParallel, false).first
                return GTLRecipeModifiers.reduction(lab, recipe,
                    lab.reductionEUt, lab.reductionDuration)
            }

        override fun onRecipeFinish() {
            machine.afterWorking()
            lastRecipe?.let { handleRecipeOutput(this.machine, it) }
            val match = if (this.isNanCertificate) gtRecipe else this.oneRecipe
            if (match != null && matchRecipeOutput(this.machine, match)) {
                setupRecipe(match)
                return
            }
            status = Status.IDLE
            progress = 0
            duration = 0
        }
    }

    companion object {
        private val RHENIUM_NANOSWARM: ItemStack = getItemStack("gtceu:rhenium_nanoswarm")
        private val ORICHALCUM_NANOSWARM: ItemStack = getItemStack("gtceu:orichalcum_nanoswarm")
        private val INFUSCOLIUM_NANOSWARM: ItemStack = getItemStack("gtceu:infuscolium_nanoswarm")
        private val NAN_CERTIFICATE: ItemStack = GTItems.NAN_CERTIFICATE.asStack()

        private val BEFORE_RECIPE: BiPredicate<GTRecipe, IRecipeLogicMachine> =
            BiPredicate { recipe: GTRecipe, machine: IRecipeLogicMachine ->
                val lab = machine as? BiologicalSimulationLaboratory ?: return@BiPredicate false
                val input = RecipeHelper.getInputItems(recipe)
                for (stack in input) {
                    if (stack.item == getItem("avaritia:infinity_sword") && !lab.isMultiRecipeEnabled()) {
                        RecipeResult.of(machine, RecipeResult.fail(
                            Component.translatable("gtceu.machine.biological_simulation_laboratory.recipe.tooltip.0")))
                        return@BiPredicate false
                    }
                }
                true
            }

        val MANAGED_FIELD_HOLDER: ManagedFieldHolder =
            ManagedFieldHolder(BiologicalSimulationLaboratory::class.java, GTLAddWorkableElectricMultipleRecipesMachine.MANAGED_FIELD_HOLDER)
    }
}
